src/twisted/mail/smtp.py

```diff
@@ -1888,14 +1898,14 @@ def __init__(self, fromEmail, toEmail, file, deferred, retries=5,
         assert isinstance(retries, (int, long))
 
         if isinstance(toEmail, unicode):
-            toEmail = [toEmail.encode('ascii')]
+            toEmail = [toEmail.encode('utf-8')]
```

Comment:
This also violates RFC 6531, because a  [a client may send UTF-8 encoded sequences if and only if an ESMTP server advertises the `SMTPUTF8` extension](https://tools.ietf.org/html/rfc6531#section-3.2).

This implementation has both `SMTPSender` and `ESMTPSender` *always* send UTF-8 encoded sequences, which violates the RFC.  `SMTPSender` should never send UTF-8, while `ESMTPSender` should only send UTF-8 when [`ESMTP.esmtpState_serverConfig` encounters the `SMTPUTF8` extension](https://github.com/twisted/twisted/blob/a8f654d642b6a2a88524dfe690a6a04814752506/src/twisted/mail/smtp.py#L1429-L1443).  `ESMTPSenderFactory` can then encode its addresses as UTF-8, and raise an informative exception if the server doesn't advertise `SMTPUTF8`.

Reply:
Third parties such as buildbot are creating [ESMTPSenderFactory](https://github.com/twisted/twisted/blob/trunk/src/twisted/mail/smtp.py#L2055) directly, which in turn calls [SMTPSenderFactory.__init__()](https://github.com/twisted/twisted/blob/trunk/src/twisted/mail/smtp.py#L1867).
`SMTPSenderFactory.__init__()` is encoding the addresses.  This is all done before the `EHLO` message is sent to the server in order query the ESMTP extensions, such as SMTPUTF8.  So SMTPSenderFactory is encoding the addresses before it receives SMTPUTF8.

Please clarify in the code how we can get this to work.
Thanks.

Reply:
@rodrigc 
```python
class SMTPSenderFactory(...):
     ..
    _encoding = 'utf-8'
    ...
    def __init__(self, ...):
        ...
        if isinstance(toEmail, unicode):
            toEmail = [toEmail.encode(self._encoding)]
        elif isinstance(toEmail, bytes):
            toEmail = [toEmail]
        else:
            toEmailFinal = []
            for _email in toEmail:
                if not isinstance(_email, bytes):
                    _email = _email.encode(self._encoding)

                toEmailFinal.append(_email)
            toEmail = toEmailFinal
        ...
```

```python
class ESMTPSenderFactory(SMTPSenderFactory):
    ...
    _encoding = 'utf-8'
```

```python
class ESMTPClient(SMTPClient):
    ...
    def esmtpState_serverConfig(self, code, resp):
        ...
        if b'SMTPUTF8' not in items:
            self.sendError(UTF8Required(...))
```

SMTP clients still encode addresses as ASCII.  ESMTP clients encode them preemptively as UTF-8, but if the server can't accept UTF-8 addresses, the Sender Factory's results Deferred errbacks with a new exception `UTF8Required`.  This gives clients an informative error message and shuts down the connection without violating the RFC and causing potentially undefined behavior.

Reply:
In your example, did you mean to put:

```
class SMTPSenderFactory(...):
     ..
    _encoding = 'ascii'
    ...
```

?

Reply:
For the case of ESMTP, if the ESMTPClient encodes addresses as unicode,
but they fall in the ascii range, should the client still throw an exception, or should
it just change the encoding to ascii, and re-encode and send it along?

Reply:
@rodrigc Whoops, yeah, I meant UTF-8 for `SMTPSenderFactory`.  It might be neat to try to re-encode the address as ASCII, but it seems kind of tricky and would be worth a log message certainly.

It'd be better to defer encoding until the extensions have come in but it will require a fair amount of refactoring.

Reply:
Also, if b'SMTPUTF8' is *not* supported by the server, would the exception be bettter named something like *UTF8NotSupported*?

Reply:
The concern I have is that for people who are now using ESMTPClient with ascii addresses against
older servers which do not support the SMTPUTF8 option, then things go through.  Throwing this exception will be new unexpected behavior.

Reply:
Maybe what I can do is modify the ESMTPClient so that if SMTPUTF8
is not in the options on the server, it inspects the bytes of the addresses,
and if any non-ASCII characters are in the address, it throws an exception.

Reply:
Note that In the case of Buildbot we want to anyway send that email, even if we have to replace non-ascii characters by ascii equivalent. Probably most of the client would have same need, but some other want to error out i that case, or at least warn in a UI. So we should probably raise that exception, and document an easy and portable way to translate that email into somthing that can be send via that old smtp server